contracts/Stake/RewardDelegators.sol

```diff
@@ -156,122 +147,108 @@ contract RewardDelegators is Initializable, Ownable {
                                                                 );
             }
         }
-        clusters[_cluster].lastRewardDistNonce = cluster.lastRewardDistNonce.add(1);
         transferRewards(clusterRegistry.getRewardAddress(_cluster), commissionReward);
         emit ClusterRewardDistributed(_cluster);
     }
 
     function delegate(
-        address _delegator, 
-        address _cluster, 
-        bytes32[] memory _tokens, 
+        address _delegator,
+        address _cluster,
+        bytes32[] memory _tokens,
         uint256[] memory _amounts
     ) public onlyStake {
         _updateRewards(_cluster);
-        Cluster memory clusterData = clusters[_cluster];
-        require(
-            clusterRegistry.isClusterValid(_cluster),
-            "ClusterRegistry:delegate - Cluster should be registered to delegate"
-        );
-        uint256 currentNonce = clusterData.lastRewardDistNonce;
-        uint256 totalRewards;
-        uint256 totalRewardDebt;
-        for(uint256 i=0; i < _tokens.length; i++) {
-            uint256 tokenAccRewardPerShare = clusters[_cluster].accRewardPerShare[_tokens[i]];
-            uint256 delegatorTokens = clusters[_cluster].delegators[_delegator][_tokens[i]];
+
+        uint256 _pendingRewards;
+        for(uint256 i = 0; i < _tokens.length; i++) {
+            uint256 _amount = _amounts[i];
+            bytes32 _tokenId = _tokens[i];
+            if(_amount == 0) continue;
+
+            uint256 _accRewardPerShare = clusters[_cluster].accRewardPerShare[_tokenId];
+            uint256 _balance = clusters[_cluster].delegators[_delegator][_tokenId];
+            uint256 _rewardDebt = clusters[_cluster].rewardDebt[_delegator][_tokenId];
 
             // calculating pending rewards for the delegator if any
-            if(clusters[_cluster].lastDelegatorRewardDistNonce[_delegator][_tokens[i]] < currentNonce) {
-                totalRewards = totalRewards.add(delegatorTokens.mul(tokenAccRewardPerShare));
-                totalRewardDebt = totalRewardDebt.add(clusters[_cluster].rewardDebt[_delegator][_tokens[i]]);
-                clusters[_cluster].lastDelegatorRewardDistNonce[_delegator][_tokens[i]] = currentNonce;
-            }
+            _pendingRewards = _pendingRewards.add(_accRewardPerShare.mul(_balance).div(10**30)).sub(_rewardDebt);
```

Comment:
Calculating totalRewards and totalRewardDebt will slightly increase the accuracy.

Reply:
true, but its also more gas though

idk, take a call

Reply:
arithmetic wise it is almost same, later approach actually becomes cheaper(tokens > 1) as tokens increase(for redelegation). But yeah 2 extra memory variables.

Anyways, the gas cost diff is too low (probably < 10). So preferring accuracy.